pkg/scheduler/queue/multi_queuing_algorithm_tree_queue_benchmark_test.go

```diff
@@ -0,0 +1,533 @@
+		)
+	}
+	return fmt.Sprintf(
+		"seconds in queue: %v", queryComponentReports,
+	)
+}
+
+func (r *testScenarioQueueDurationReport) TenantIDReportString() string {
+	var tenantIDReports []string
+	for _, tenantIDReport := range r.tenantIDReports {
+		tenantIDReports = append(
+			tenantIDReports,
+			fmt.Sprintf(
+				"%s: mean: %.4f stddev: %.2f",
+				tenantIDReport.dimensionName,
+				tenantIDReport.meanSecondsInQueue,
+				tenantIDReport.stdDevSecondsInQueue),
+		)
+	}
+	return fmt.Sprintf(
+		"seconds in queue:%v", tenantIDReports,
+	)
+}
+
+type testDimensionReport struct {
+	dimensionName        string
+	meanSecondsInQueue   float64
+	stdDevSecondsInQueue float64
+}
+
+func mean(numbers []float64) float64 {
+	sum := 0.0
+	for _, number := range numbers {
+		sum += number
+	}
+	return sum / float64(len(numbers))
+}
+
+func stddev(numbers []float64, mean float64) float64 {
+	sumOfSquares := 0.0
+	for _, number := range numbers {
+		sumOfSquares += math.Pow(number-mean, 2)
+	}
+	meanOfSquares := sumOfSquares / float64(len(numbers))
+	return math.Sqrt(meanOfSquares)
+}
+
+// TestMultiDimensionalQueueFairnessSlowConsumerEffects emulates a simplified queue slowdown scenario
+// which the scheduler's additional queue dimensions features are intended to solve for.
+//
+// In this scenario, one category of queue item causes the queue consumer to slow down, introducing a
+// significant delay while the queue consumer processes it and before the consumer can dequeue the next item.
+// This emulates a situation where one of the query components - the ingesters or store-gateways - is under load.
+//
+// In a single-queue or single-queue-per-tenant scenario, the slow query component will cause the queue to back up.
+// When queue items belonging to the slow category are in the same queue in front of the normal queue items,
+// the normal queue items must wait for all slow queue items to be cleared before they can be serviced.
+// In this way, the degraded performance of the slow query component equally degrades the performance of the
+// queries which *could* be serviced quickly, but are waiting behind the slow queries in the queue.
+//
+// In this benchmark, queries for the store-gateway query component are artificially slowed down.,
+// We assume that in such a slowdown scenario that queries hitting both query components (ingester and store-gateway)
+// will be constrained by the latency of the slowest query component, so we treat them as store-gateway queries.
+// The goal of the queueing algorithms is to enable ingester-only queries to continue to be serviced quickly
+// when the store-gateway queries are slow; we are not concerned with de-loading the store-gateways
+// or otherwise improving the performance for the store-gateway queries.
+// Benchmark results should be interpreted primarily in terms of the performance of the ingester-only queries.
+func TestMultiDimensionalQueueAlgorithmSlowConsumerEffects(t *testing.T) {
```

Comment:
IIUC there's no assertion in this test to ensure that we perform better/worse with specific test configurations; is there a way we can add this, but not have it run with the general unit tests? Running the tests in CI for the `queue` package went from taking ~30s to ~172s with this change.

Reply:
I wanna check with some other people on how they feel about this - I got some pushback on asserting on the benchmark outputs on the previous version of this for fear of them being flaky in CI, but I was able to configure it where it was stable.

I think there's also a fair question of like, when this is all said and done after the cleanup we won't be using vanilla round robin anywhere. Should we still keep something like that in a code base just to show it's worse?

w.r.t the total test runtime, we have a few levers:
* remove one of the two slow tree configurations
* lower the iterations
* lower the slow consumer delay
* crank up worker concurrency
* nuclear: skip in CI altogether somehow until we come up with a larger strategy for regular benchmarking

Reply:
so I lopped off the 5% / 95%  cases since they generally show the same effects as the 10% / 90% effects and these cases take the longest
I also cranked up querier and worker parallellism - total runtime on my local is down to 14s.

As far as asserting on differences, it looks like I could probably assert on the new algo being 10x better in all cases _except_ the 10% case. In general with lower % slow queries, the gains are less pronounced so it would need to be a sliding scale. Gonna ask some more mimir people how they generally feel about this sort of approach.